### DISCUSSION TITLE
Texture format for indirect light

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#3520] [Q&A] [answered]
I tried using `TextureFormat::RGBA8UI` for a cube map texture to be used in an `IndirectLight`, and got an assertion failure:

```
Fragment Function(main0): The pixel format (MTLPixelFormatRGBA8Uint) of the texture (name:<null>) bound at index 4 is incompatible with the data type (MTLDataTypeFloat) of the texture parameter (light_iblSpecular [[texture(0)]]). MTLPixelFormatRGBA8Uint is compatible with the data type(s) (
    uint,
    ushort
).
```

Does an `IndirectLight` always require a floating-point texture format, or does that depend on the platform?

**Comment by romainguy** [comment, marked as answer]
Metal has more constraints but you should use a float format anyway. IBLs encode HDR lighting after all. Just use the R11G11B10F format, it’s still 32 bit per pixel but contains float data. If your IBL is small you can use RGBA16F as it’s easier to encore (you can use our half.h API for that).

**Comment by prideout** [comment]
By the way are you using cmgen to generate the IBL?  I'm surprised it can even generate textures in that format.

**Reply by romainguy** [reply]
It doesn’t, cmgen only generates R11G11B10F or RGBM.

**Reply by jwwalker** [reply]
That's a fair point; `IndirectLight::Builder::reflections` does say that the cube map should be generated by `cmgen`.  I just thought that when I wanted to make an all-white cube map, it would be easier to make one from scratch rather than figure out how to use `cmgen`.  Maybe that was a miscalculation on my part.

**Reply by romainguy** [reply]
You don't have to use cmgen, but on Metal you need to make sure the data is in a float format.
